ocaml/testsuite/tests/pattern-guards/test.ml

```diff
@@ -260,3 +260,55 @@ say_hello_catching_exns None name_map;;
 [%%expect{|
 - : string = "Hello, Fred"
 |}]
+
+let single_bar_syntax x =
+  match x with
+  | Some x when x match | Some y -> y
+  | _ -> 0
+;;
+[%%expect{|
+val single_bar_syntax : int option option -> int = <fun>
+|}];;
+
+single_bar_syntax (Some (Some 5));;
+[%%expect{|
+- : int = 5
+|}];;
+single_bar_syntax (Some None);;
+[%%expect{|
+- : int = 0
+|}];;
+single_bar_syntax None;;
+[%%expect{|
+- : int = 0
+|}];;
+
+let find_multiway xs = function
+  | x :: _ when List.find xs x match (
+      | Some true -> 1
+      | Some false -> 0
+    )
+  | _ -> ~-1
+;;
+[%%expect{|
+>> Fatal error: typechecking for multicase pattern guards unimplemented
+Uncaught exception: Misc.Fatal_error
+
+|}]
+
+let nested_multiway f g h = function
+  | Some x when f x match (
+      | "foo" when h x -> "foo1"
+      | "bar" when g x match (
+          | [] -> "bar empty"
+          | [ y ] -> "bar singleton " ^ y
+        )
+    )
+  | _ -> "not found"
+;;
+[%%expect{|
+>> Fatal error: typechecking for multicase pattern guards unimplemented
+Uncaught exception: Misc.Fatal_error
+
+|}]
+
```

Comment:
How about nested single-way matches? I don't think those will typecheck yet, but it will exercise the parser.

Reply:
Added! This helped me catch a bug in parmatch where we incorrectly issue the warning `All_clauses_guarded` for clauses in a pattern guard.